### PR TITLE
HADOOP-16680. Add MicrosoftGraphGroupsMapping GroupMappingServiceProvider

### PR DESCRIPTION
**Pull request by lukmajercak**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 33 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1232 | trunk passed |
| +1 | compile | 1058 | trunk passed |
| +1 | checkstyle | 44 | trunk passed |
| +1 | mvnsite | 81 | trunk passed |
| +1 | shadedclient | 943 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 86 | trunk passed |
| 0 | spotbugs | 126 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 123 | trunk passed |
| -0 | patch | 157 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 57 | the patch passed |
| +1 | compile | 1040 | the patch passed |
| -1 | javac | 1040 | root generated 2 new + 1888 unchanged - 0 fixed = 1890 total (was 1888) |
| -0 | checkstyle | 46 | hadoop-common-project/hadoop-common: The patch generated 12 new + 0 unchanged - 0 fixed = 12 total (was 0) |
| +1 | mvnsite | 77 | the patch passed |
| +1 | whitespace | 1 | The patch has no whitespace issues. |
| +1 | xml | 1 | The patch has no ill-formed XML file. |
| -1 | shadedclient | 176 | patch has errors when building and testing our client artifacts. |
| +1 | javadoc | 84 | the patch passed |
| -1 | findbugs | 135 | hadoop-common-project/hadoop-common generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0) |
||| _ Other Tests _ |
| -1 | unit | 534 | hadoop-common in the patch failed. |
| -1 | asflicense | 45 | The patch generated 11 ASF License warnings. |
| | | 5875 | |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-common-project/hadoop-common |
|  |  Found reliance on default encoding in org.apache.hadoop.security.oauth2.AzureADAuthenticator.consumeInputStream(InputStream):in org.apache.hadoop.security.oauth2.AzureADAuthenticator.consumeInputStream(InputStream): new String(byte[], int, int)  At AzureADAuthenticator.java:[line 153] |
|  |  org.apache.hadoop.security.oauth2.QueryParams.serialize() makes inefficient use of keySet iterator instead of entrySet iterator  At QueryParams.java:keySet iterator instead of entrySet iterator  At QueryParams.java:[line 44] |
| Failed junit tests | hadoop.conf.TestCommonConfigurationFields |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1690/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1690 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient xml findbugs checkstyle |
| uname | Linux 6de7a5250d7e 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 7d7acb0 |
| Default Java | 1.8.0_222 |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1690/1/artifact/out/diff-compile-javac-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1690/1/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1690/1/artifact/out/new-findbugs-hadoop-common-project_hadoop-common.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1690/1/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1690/1/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-1690/1/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 1442 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1690/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
I really don't want add new dependencies here unless critical.

1. Can it go into hadop-azure?
1. And be tagged as private?

I don't expect there to be any deployments of this mapping except in installations with hadoop-azure on the classpath. Putting it there would line things up as "part of the Azure code" and avoid problems for everyone else.

**Comment by lukmajercak** [comment]
@steveloughran not sure what you mean by "installations with hadoop-azure" on the classpath. Afaik all recent hadoop installations will have that. This is also not azure specific code. We run hadoop on a non-azure environment but we use msgraph for authZ.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/MicrosoftGraphGroupsMapping.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
+package org.apache.hadoop.security;
+
+import com.google.common.annotations.VisibleForTesting;
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/MicrosoftGraphGroupsMapping.java]: import ordering

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/MicrosoftGraphGroupsMapping.java [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+ * token to retrieve a list of groups of the requested user.
+ *
+ */
+public class MicrosoftGraphGroupsMapping
+    implements GroupMappingServiceProvider, Configurable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      MicrosoftGraphGroupsMapping.class);
+
+  public static final String MS_GRAPH_GROUPS_OAUTH2_URL_KEY =
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/MicrosoftGraphGroupsMapping.java]: javadocs of all the constants

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/MicrosoftGraphGroupsMapping.java [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    // Get the client secret
+    String oauth2ClientSecret = "";
+    try {
+      char[] clientSecret = conf.getPassword(oauth2ClientSecretAlias);
+      if (clientSecret != null) {
+        oauth2ClientSecret = new String(clientSecret);
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to get clientSecret using alias: {}",
+          oauth2ClientSecretAlias);
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/MicrosoftGraphGroupsMapping.java]: include the exception

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/MicrosoftGraphGroupsMapping.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+   * Microsoft Graph API to get the groups.
+   *
+   * @param user get groups for this user
+   * @return list of groups for a given user
+   */
+  @Override // GroupMappingServiceProvider
+  public List<String> getGroups(final String user) throws IOException {
+
+    // TODO: This is a hack to fail fast for 'hadoop' username
+    // to avoid API calls and retries. Should be removed later.
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/MicrosoftGraphGroupsMapping.java]: "should be removed later"?

**Comment by lukmajercak** [comment]
Doesn't spark use hadoop-common for authz though? So if we wanna use msgraph for authz in spark we'll have to add hadoop-azure as a dependency?

**Comment by steveloughran** [comment]
if you build spark with hadoop-cloud-storage enabled you get hadoop-azure and dependencies. And if you build your own hadoop release you can patch whatever you want into hadoop-cloud-storage and see it picked up.

Now, to repeat: I am going to -1 all changes which add yet another JAR dependency tree for every single use of hadoop-common unless there is absolutely no choice whatsoever in core code. I Don't see that here, and everything else azure related gets to live in hadoop-azure.

Sorry. We cause enough classpath problems downstream to add more. HBase still haven't spoken to me since I updated Jackson in 2013.

Side issue: would people want to make this new JAR and dependencies a mandatory or optional dependency of hadoop-azure? That I delegate to @ThomasMarquardt and @DadanielZ

**Comment by goiri** [comment]
@steveloughran I created a separate PR moving it to hadoop-tools:
https://github.com/apache/hadoop/pull/1727

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m  0s |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  5s |  https://github.com/apache/hadoop/pull/1690 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/1690 |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1690/3/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m  0s |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  5s |  https://github.com/apache/hadoop/pull/1690 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/1690 |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1690/4/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m  0s |  |  Docker mode activated.  |
| -1 :x: |  patch  |   0m 16s |  |  https://github.com/apache/hadoop/pull/1690 does not apply to trunk. Rebase required? Wrong Branch? See https://cwiki.apache.org/confluence/display/HADOOP/How+To+Contribute for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/1690 |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-1690/1/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
closing as it is superceded by #1727. Let's see if we can get it wrapped up in there

**Pull request closed without merging**